// @kintone/plugin-uploader/dist/pages/OldPluginSystemPage.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OldPluginSystemPage = exports.IMPORT_BUTTON_SELECTOR = void 0;
const messages_1 = require("../messages");
const chalk_1 = __importDefault(require("chalk"));
const TIMEOUT_MS = 10000;
const UPLOAD_TIMEOUT_MS = 60000;
exports.IMPORT_BUTTON_SELECTOR = "#page-admin-system-plugin-index-addplugin";
const IMPORT_PLUGIN_DIALOG_SELECTOR = ".ocean-ui-dialog";
const FILE_SELECTOR = '.plupload > input[type="file"]';
class OldPluginSystemPage {
    readyForImportButton(page, boundMessage) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield page.waitForSelector(exports.IMPORT_BUTTON_SELECTOR, {
                    timeout: TIMEOUT_MS,
                });
            }
            catch (e) {
                throw chalk_1.default.red(boundMessage("Error_notDisplayImportButton"));
            }
        });
    }
    upload(page, pluginPath, lang) {
        return __awaiter(this, void 0, void 0, function* () {
            const boundMessage = (0, messages_1.getBoundMessage)(lang);
            console.log(`Trying to upload ${pluginPath}`);
            yield page.click(exports.IMPORT_BUTTON_SELECTOR);
            yield page.waitForSelector(IMPORT_PLUGIN_DIALOG_SELECTOR, {
                timeout: TIMEOUT_MS,
            });
            const file = yield page.$(FILE_SELECTOR);
            if (file === null) {
                throw new Error('input[type="file"] is not found');
            }
            yield file.uploadFile(pluginPath);
            // HACK: `page.click` does not work as expected, so we use `page.evaluate` instead.
            // ref: https://github.com/puppeteer/puppeteer/pull/7097#issuecomment-850348366
            yield page.evaluate(() => {
                const button = document.querySelector('button[name="ok"]');
                if (button) {
                    button.click();
                }
                else {
                    throw new Error('button[name="ok"] is not found');
                }
            });
            yield page.waitForSelector(".ocean-ui-dialog", {
                hidden: true,
                timeout: UPLOAD_TIMEOUT_MS,
            });
            console.log(`${pluginPath} ${boundMessage("Uploaded")}`);
        });
    }
}
exports.OldPluginSystemPage = OldPluginSystemPage;
